### PR TITLE
CI: Use local cache registry

### PR DESCRIPTION
**Pull request by edsantiago**
As of https://github.com/containers/automation_images/pull/357 our CI VMs include a local registry preloaded with all(*) images used in tests.

(*) where "all" means "most".

This PR sets up CI such that it will use that registry.

Signed-off-by: Ed Santiago <santiago@redhat.com>
```release-note
None
```

### CODE REVIEW

File: test/e2e/pull_test.go
Comment (Luap99) [inline test/e2e/pull_test.go:42]: heads up, tests should still pass locally and I don't think we want to setup a local registry there right?
Thus I would think we need a regex or a `Or()` matcher to match both strings anyway

File: contrib/cirrus/runner.sh
Comment (Luap99) [inline contrib/cirrus/runner.sh:149]: this does not do anything, not sure if you were expecting anything with that or if this serves documentation purposes.

Comment (edsantiago) [reply contrib/cirrus/runner.sh:149]: I need the container to be able to talk to host:56789. `--port` does the opposite, IIRC: host can talk to container. I'll look at logs and see what happens.

File: test/e2e/pull_test.go
Comment (edsantiago) [reply test/e2e/pull_test.go:42]: This is thorny. As tests are currently written, e2e tests hard-force the use of `test/registries.conf`. No matter where they're run (CI, laptop, anywhere). This may need to be reevaluated, but I'm not bothering with any of that until I find out if this approach is viable.

File: contrib/cirrus/runner.sh
Comment (Luap99) [reply contrib/cirrus/runner.sh:149]: The container uses --network=host so it shares the network namespace with the host so from a network POV there should be no functional difference

File: test/e2e/pull_test.go
Comment (Luap99) [reply test/e2e/pull_test.go:42]: yes sure keep testing, just keep in mind that the end result must still work locally

File: contrib/cirrus/runner.sh
Comment (edsantiago) [reply contrib/cirrus/runner.sh:149]: Duh. Thanks, I missed that.

File: contrib/cirrus/local-cache-registry
Comment (Luap99) [inline contrib/cirrus/local-cache-registry:251]: I recommend to use `--network host` and set `-e REGISTRY_HTTP_ADDR=0.0.0.0:${PODMAN_REGISTRY_PORT}` instead.
Reason by using already a podman container you sort of conflict with later running containers networking wise. The network reload tests that nuke iptables,etc... will cause issues for connectivity one way or another.

Comment (edsantiago) [reply contrib/cirrus/local-cache-registry:251]: ohhhhhh.. thank you. That's probably the cause of the hangs in the 700-kube tests.

**Comment by edsantiago** [comment]
@containers/podman-maintainers this is getting close. There's one question I can't decide and would like help with: how to handle local developer-workstation runs:
1. same as CI: force devs to set up and run a local registry. This is what my third commit is intended for. Or:
2. same as how things are today: local-dev runs will use actual real-world registries.

Both options stink. Option 1 imposes work and complexity on anyone running tests. Option 2 would require maintenance of almost-duplicate `registries.conf` files, and a mechanism for tests to decide which to use when, and conditionals for a handful of tests that behave differently (403 vs 404 on no-such-image pulls); and of course local-environment runs will not be testing the same as CI.

As it stands right now, this PR implements option 1. Your feedback would be most welcome.

**Comment by Luap99** [comment]
> 2. same as how things are today: local-dev runs will use actual real-world registries.

I prefer 2 but can be convinced that 1 is better. However if a local registry is required then IMO it must be part of the test setup and not user responsibility. If I run `test/apiv2/test-apiv2|make localintegration|hack/bats` today that must keep working and I should not have to run another script manually before that. Given that I do not see how this would deal with pulling images for the local registry? I do not want to wait a very long time pulling images the first time when I run a system test which only ever needs one tiny image.

**Comment by edsantiago** [comment]
>if a local registry is required then IMO it must be part of the test setup and not user responsibility

Fair point, and that's close to impossible. And if it _were_ possible it would be unbearable to wait 20-60m for the cache to preload. Okay, Option 1 it is then.

**Comment by edsantiago** [comment]
This implements option 1: tests in CI will use the cache registry, but anyone running tests on a laptop will still get the old behavior (pulling from real registries). The switch is `$CI_USE_REGISTRY_CACHE`

**Comment by edsantiago** [comment]
If this is acceptable, I think it just needs documentation, and one more CI VM build

**Review by Luap99 — COMMENTED** [review]
Not a full review yet but assume commit 3 is only interesting in case of CI_USE_REGISTRY_CACHE set locally given we decided to not require the registry script?
If that is the case can you update the commit and add this as a comment in the fetch script.

**Review by Luap99 — COMMENTED** [review]
LGTM

Thinking about this should we set up a new cron run that runs without the cache registry to make sure local tests will keep working? (not blocking of course)

**Comment by edsantiago** [comment]
Good idea. 

/hold

Let's not merge this yet. I saw [this bud flake](https://api.cirrus-ci.com/v1/artifact/task/4673065106276352/html/bud-podman-fedora-40-root-host-sqlite.log.html#t--00319) over the weekend and I want to understand it better.

File: test/e2e/info_test.go
Comment (edsantiago) [inline test/e2e/info_test.go:124]: This is, I think, the only change since last review. The latest Debian VM image creates and recreates `/run/user/XXX/podman/podman.sock` at (seemingly) random times. I've given up trying to understand what is creating it, and honestly I choose to decide that this test was broken. There is no harm in the socket _existing_ in a podman-local test. The only necessary test is that if podman-remote, then socket-must-exist.

**Review by edsantiago — COMMENTED** [review]
Once again I think this is ready for review. My thanks for everyone's patience.

**Review by Luap99 — APPROVED** [review]
LGTM

**Comment by rhatdan** [comment]
/lgtm
I will let @edsantiago cancel the hold

**Comment by edsantiago** [comment]
/hold cancel

(I forgot about that). Thanks.

**Pull request merged**
